src/emit/binary.rs

```diff
@@ -805,33 +805,23 @@ impl<'a> Binary<'a> {
                     self.module.get_struct_type("struct.vector").unwrap().into()
                 }
                 Type::Array(base_ty, dims) => {
-                    let ty = self.llvm_field_ty(base_ty, ns);
-
-                    let mut dims = dims.iter();
-
-                    let mut aty = match dims.next().unwrap() {
-                        ArrayLength::Fixed(d) => ty.array_type(d.to_u32().unwrap()),
-                        ArrayLength::Dynamic => {
-                            return self.module.get_struct_type("struct.vector").unwrap().into()
-                        }
-                        ArrayLength::AnyFixed => {
-                            unreachable!()
-                        }
-                    };
+                    let mut aty = self.llvm_field_ty(base_ty, ns);
 
                     for dim in dims {
                         match dim {
-                            ArrayLength::Fixed(d) => aty = aty.array_type(d.to_u32().unwrap()),
+                            ArrayLength::Fixed(d) => {
+                                aty = aty.array_type(d.to_u32().unwrap()).into();
+                            }
                             ArrayLength::Dynamic => {
-                                return self.module.get_struct_type("struct.vector").unwrap().into()
+                                aty = self.module.get_struct_type("struct.vector").unwrap().into();
                             }
                             ArrayLength::AnyFixed => {
                                 unreachable!()
                             }
                         }
                     }
 
-                    BasicTypeEnum::ArrayType(aty)
+                    aty
```

Comment:
Can be simplified (unless the loop introduces side-effects, which would rather surprise me):  

```rust
match dims.last() {
  Some(ArrayLength::Fixed(d)) => self.llvm_field_ty(base_ty, ns).array_type(d.to_u32().unwrap()).into(),
  Some(ArrayLength::Dynamic) => self.module.get_struct_type("struct.vector").unwrap().into(),
  None => self.llvm_field_ty(base_ty, ns),
  _ => unreachable!()
}
```

Reply:
We want to walk all the dimensions (e.g. `int32[4][3][2]` needs to be the same in llvm). However you do have a point, it could be simpler. I've re-written it using `.fold()`. Hope this is better - it's certainly shorter.

Reply:
Ah I see. Yes that's much shorter and quicker to grasp.

However, I think I overlooked the early return in case of` ArrayLength::Dynamic` in the original version. That might actually play a role here?

Reply:
Very well then